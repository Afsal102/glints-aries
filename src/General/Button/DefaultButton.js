/* @flow */

import React from 'react';
import { DefaultBtnContainer, DefaultBtn } from '../../Style/General/ButtonStyle';

const DefaultButton = ({
  children,
  className,
  theme,
  disabled,
  block,
  small,
  removeHoverEffect,
  ...defaultProps
}: Props) => (
  <DefaultBtnContainer
    id="aries-defaultbtn"
    className={className}
    theme={theme}
    disabled={disabled}
    block={block}
    small={small}
    removeHoverEffect={removeHoverEffect}
  >
    <DefaultBtn
      id="defaultbtn-content"
      theme={theme}
      disabled={disabled}
      block={block}
      small={small}
      {...defaultProps}
    >
      {children}
    </DefaultBtn>
  </DefaultBtnContainer>
);

type Props = {
  children: React$Node,
  className: string,
  theme: string,
  disabled: boolean,
  block: boolean,
  small: boolean,
}

export default DefaultButton;
