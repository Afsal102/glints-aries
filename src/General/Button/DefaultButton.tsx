import * as React from 'react';
import { DefaultBtnContainer, DefaultBtn } from '../../Style/General/ButtonStyle';

const DefaultButton: React.FunctionComponent<Props> = ({
  children,
  className,
  theme,
  disabled,
  block,
  small,
  removeHoverEffect,
  ...defaultProps
}) => (
  <DefaultBtnContainer
    id="aries-defaultbtn"
    className={className}
    theme={theme}
    disabled={disabled}
    block={block}
    removeHoverEffect={removeHoverEffect}
  >
    <DefaultBtn
      id="defaultbtn-content"
      theme={theme}
      disabled={disabled}
      block={block}
      small={small}
      {...defaultProps}
    >
      {children}
    </DefaultBtn>
  </DefaultBtnContainer>
);

interface Props extends React.ComponentPropsWithoutRef<typeof DefaultBtn> {
  children: React.ReactNode;
  className: string;
  theme: string;
  disabled?: boolean;
  block: boolean;
  small: boolean;
  removeHoverEffect: boolean;
}

export default DefaultButton;
