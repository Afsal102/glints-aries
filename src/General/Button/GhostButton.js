/* @flow */

import React from 'react';
import { GhostBtnContainer, GhostBtn } from '../../Style/General/ButtonStyle';

const GhostButton = ({
  children,
  className,
  theme,
  disabled,
  block,
  small,
  removeHoverEffect,
  ...defaultProps
}: Props) => (
  <GhostBtnContainer
    id="aries-ghostbtn"
    className={className}
    theme={theme}
    disabled={disabled}
    block={block}
    small={small}
    removeHoverEffect={removeHoverEffect}
  >
    <GhostBtn
      id="ghostbtn-content"
      theme={theme}
      disabled={disabled}
      block={block}
      small={small}
      {...defaultProps}
    >
      {children}
    </GhostBtn>
  </GhostBtnContainer>
);

type Props = {
  children: React$Node,
  className: string,
  theme: string,
  disabled: boolean,
  block: boolean,
  small: boolean,
}

export default GhostButton;
