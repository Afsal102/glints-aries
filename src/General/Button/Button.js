/* @flow */

import React, { Fragment } from 'react';

import DefaultButton from './DefaultButton';
import PrimaryButton from './PrimaryButton';
import SecondaryButton from './SecondaryButton';
import GhostButton from './GhostButton';
import LinkButton from './LinkButton';

import { Variant } from '../../Utils/StyleConfig';

const renderButton = ({
  variant,
  theme,
  children,
  onClick,
  className,
  ...defaultProps
}: Props) => {
  switch (variant) {
    case Variant.PRIMARY:
      return (
        <PrimaryButton
          theme={theme}
          className={className}
          onClick={onClick}
          {...defaultProps}
        >
          {children}
        </PrimaryButton>
      );
    case Variant.SECONDARY:
      return (
        <SecondaryButton
          className={className}
          onClick={onClick}
          {...defaultProps}
        >
          {children}
        </SecondaryButton>
      );
    case Variant.GHOST:
      return (
        <GhostButton
          theme={theme}
          className={className}
          onClick={onClick}
          {...defaultProps}
        >
          {children}
        </GhostButton>
      );
    case Variant.LINK:
      return (
        <LinkButton
          theme={theme}
          className={className}
          onClick={onClick}
          {...defaultProps}
        >
          {children}
        </LinkButton>
      );
    default:
      return (
        <DefaultButton
          theme={theme}
          className={className}
          onClick={onClick}
          {...defaultProps}
        >
          {children}
        </DefaultButton>
      );
  }
};

const Button = (props: Props) => (
  <Fragment>
    {renderButton(props)}
  </Fragment>
);

type Props = {
  variant: string,
  theme: string,
  onClick: Function,
  children: React$Node,
  className: string,
}

export default Button;
