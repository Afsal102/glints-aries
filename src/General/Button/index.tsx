import Button, { DeprecatedSecondayVariant } from './Button';

export { Button, DeprecatedSecondayVariant };

export default Button;
