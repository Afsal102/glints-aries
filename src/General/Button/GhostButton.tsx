import * as React from 'react';
import classNames from 'classnames';
import { GhostBtnContainer, GhostBtn } from '../../Style/General/ButtonStyle';

const GhostButton: React.FunctionComponent<Props> = ({
  children,
  className,
  theme,
  disabled,
  block,
  small,
  removeHoverEffect,
  tag,
  ...defaultProps
}) => (
  <GhostBtnContainer
    className={classNames('aries-ghostbtn', className)}
    theme={theme}
    disabled={disabled}
    block={block}
    removeHoverEffect={removeHoverEffect}
  >
    <GhostBtn
      className="ghostbtn-content"
      theme={theme}
      disabled={disabled}
      block={block}
      small={small}
      as={(tag as React.ElementType) || 'button'}
      {...defaultProps}
    >
      {children}
    </GhostBtn>
  </GhostBtnContainer>
);

interface Props extends React.ComponentPropsWithoutRef<typeof GhostBtn> {
  children: React.ReactNode;
  className?: string;
  theme?: string;
  disabled?: boolean;
  block?: boolean;
  small?: boolean;
  removeHoverEffect?: boolean;
  tag?: React.ElementType;
}

export default GhostButton;
