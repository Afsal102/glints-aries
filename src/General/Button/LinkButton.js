/* @flow */

import React from 'react';
import { LinkBtn } from '../../Style/General/ButtonStyle';

const LinkButton = ({ children, ...defaultProps }: Props) => (
  <LinkBtn
    id="aries-linkbtn"
    {...defaultProps}
  >
    {children}
  </LinkBtn>
);

type Props = {
  children: React$Node,
  className: string,
}

export default LinkButton;
