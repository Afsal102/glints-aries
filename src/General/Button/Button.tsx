import * as React from 'react';

import SolidButton from './SolidButton';
import SolidShadowButton from './SolidShadowButton';
import GhostButton from './GhostButton';
import LinkButton from './LinkButton';

import { ButtonVariant } from '../../Utils/StyleConfig';
import {
  StartIconContainer,
  EndIconContainer,
} from '../../Style/General/ButtonStyle';

const renderButton: React.FunctionComponent<Props> = ({
  children,
  block,
  className,
  disabled,
  onClick,
  removeHoverEffect,
  small,
  theme,
  variant,
  startIcon,
  endIcon,
  ...defaultProps
}) => {
  const content = (
    <>
      {startIcon && <StartIconContainer>{startIcon}</StartIconContainer>}
      {children}
      {endIcon && <EndIconContainer>{endIcon}</EndIconContainer>}
    </>
  );

  switch (variant) {
    case ButtonVariant.SOLID_SHADOW:
      return (
        <SolidShadowButton
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          {...defaultProps}
        >
          {content}
        </SolidShadowButton>
      );
    case ButtonVariant.GHOST:
      return (
        <GhostButton
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          removeHoverEffect={removeHoverEffect}
          {...defaultProps}
        >
          {content}
        </GhostButton>
      );
    case ButtonVariant.LINK:
      return (
        <LinkButton
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          {...defaultProps}
        >
          {children}
        </LinkButton>
      );
    default:
      return (
        <SolidButton
          theme={theme}
          className={className}
          disabled={disabled}
          onClick={onClick}
          block={block}
          small={small}
          removeHoverEffect={removeHoverEffect}
          {...defaultProps}
        >
          {content}
        </SolidButton>
      );
  }
};

const Button: React.FunctionComponent<Props> = props => (
  <React.Fragment>{renderButton(props)}</React.Fragment>
);

export interface Props {
  children: React.ReactNode;
  block?: boolean;
  className?: string;
  disabled?: boolean;
  onClick?(e: React.MouseEvent<HTMLButtonElement, MouseEvent>): void;
  removeHoverEffect?: boolean;
  small?: boolean;
  theme?: string;
  variant?: string;
  tag?: React.ElementType;
  startIcon?: React.ReactNode;
  endIcon?: React.ReactNode;
}

export default Button;
