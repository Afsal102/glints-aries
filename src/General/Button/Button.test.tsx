import * as React from 'react';
import Button from '.';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

import { Variant } from '../../Utils/StyleConfig';

it('<Button> should render with text "click me" and an onClick handler', () => {
  const onClick = jest.fn();
  const ButtonSnapshot = renderer
    .create(<Button onClick={onClick}>click me</Button>)
    .toJSON();
  expect(ButtonSnapshot).toMatchSnapshot();
});

it('<Button> should render as button by default', () => {
  const tag = 'a';
  const { container } = render(<Button tag={tag}>test</Button>);
  expect(container.querySelector(tag)).toBeTruthy();
});

describe('<Button> should render as the given tag', () => {
  const tags = ['a', 'h1', 'div'];
  tags.forEach((tag: string) => {
    it(`<Button> should render as ${tag}`, () => {
      const { container } = render(
        <Button tag={tag as React.ElementType}>button</Button>
      );
      const tagRegexp = new RegExp(tag, 'i');
      expect(container.querySelector(tag).tagName).toMatch(tagRegexp);
    });
  });
});

it('should call onClick once when it is clicked', () => {
  const onClick = jest.fn();
  const { getByText } = render(<Button onClick={onClick}>click me</Button>);
  const button = getByText('click me');
  fireEvent.click(button);
  expect(onClick).toHaveBeenCalledTimes(1);
});

it('should not call onClick when it is disabled', () => {
  const onClick = jest.fn();
  const { getByText } = render(
    <Button onClick={onClick} disabled>
      click me
    </Button>
  );
  const button = getByText('click me');
  fireEvent.click(button);
  expect(onClick).toHaveBeenCalledTimes(0);
});

describe('contains the correct className for each variant:', () => {
  const variants = [...Object.values(Variant), 'default'];
  variants.forEach(variant => {
    it(`${variant}`, () => {
      const { getByText } = render(<Button variant={variant}>click me</Button>);
      if (variant === Variant.LINK) {
        const linkButton = getByText('click me');
        expect(linkButton).toHaveClass(`aries-${variant}btn`);
      } else {
        const buttonContainer = getByText('click me').parentNode;
        expect(buttonContainer).toHaveClass(`aries-${variant}btn`);
      }
    });
  });
});
