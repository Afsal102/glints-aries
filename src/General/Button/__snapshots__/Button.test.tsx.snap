// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> should render with text "click me" and an onClick handler 1`] = `
<div
  className="ButtonStyle__DefaultBtnContainer-gw4ytw-2 jOFFLF aries-defaultbtn"
>
  <button
    className="ButtonStyle__Button-gw4ytw-0 ButtonStyle__DefaultBtn-gw4ytw-1 ixCkld defaultbtn-content"
    onClick={[MockFunction]}
  >
    click me
  </button>
</div>
`;
